src/circuit/gadgets/constraints.rs

```diff
@@ -698,6 +799,172 @@ mod tests {
 
     proptest! {
 
+        #[test]
+        fn test_enforce_equal((a, b) in any::<(FWrap<Fr>, FWrap<Fr>)>()) {
+            prop_assume!(a != b);
+
+            let test_a_b = |a, b| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let a_num = AllocatedNum::alloc(cs.namespace(|| "a_num"), || Ok(a)).unwrap();
+                let b_num = AllocatedNum::alloc(cs.namespace(|| "b_num"), || Ok(b)).unwrap();
+                let _ = enforce_equal(&mut cs, || "enforce equal", &a_num, &b_num);
+                assert_eq!(cs.is_satisfied(), a==b);
+            };
+
+            // positive
+            test_a_b(a.0, a.0);
+
+            // negative
+            test_a_b(a.0, b.0);
+        }
+
+        #[test]
+        fn test_enforce_equal_zero(a in (1u64..42)) {
+
+            let test_num = |n: u64| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let num = AllocatedNum::alloc(cs.namespace(|| "zero"), || Ok(Fr::from(n))).unwrap();
+                let _ = enforce_equal_zero(&mut cs, || "enforce equal zero", &num);
+                assert_eq!(cs.is_satisfied(), n==0);
+
+            };
+
+            // positive
+            test_num(0);
+
+            // negative
+            test_num(a);
+        }
+
+        #[test]
+        fn test_implies_equal_zero(p in any::<bool>()) {
+
+            let test_premise_num = |premise: bool, n, result: bool| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let num = AllocatedNum::alloc(cs.namespace(|| "num"), || Ok(Fr::from(n))).unwrap();
+                let pb = Boolean::constant(premise);
+                let _ = implies_equal_zero(&mut cs.namespace(|| "implies equal zero"), &pb, &num);
+                assert_eq!(cs.is_satisfied(), result);
+
+            };
+
+            // any premise
+            test_premise_num(p, 0, true);
+
+            // false premise, any value
+            let rand_a = rand::thread_rng().gen_range(0..u64::MAX);
```

Comment:
Can't we make proptest give us these random numbers?

Reply:
Wouldn't it be better (if we're taking this approach) that *these* be any field element?

Reply:
done

Reply:
We would need the "non-zero-field-element" to replace u64 by field here.